src/python/pants/backend/codegen/protobuf/go/rules.py

```diff
@@ -0,0 +1,251 @@
+# copy the following go.mod and go.sum contents to go.mod and go.sum files in a new directory. Then update the
+# versions and run `go mod download all`. Copy the go.mod and go.sum contents back into these constants,
+# making sure to repalce tabs with `\t`.
+
+GO_PROTOBUF_GO_MOD = """\
+module org.pantsbuild.backend.go.protobuf
+
+go 1.17
+
+require (
+\tgoogle.golang.org/grpc/cmd/protoc-gen-go-grpc v1.2.0
+\tgoogle.golang.org/protobuf v1.27.1
+)
+
+require (
+\tgithub.com/golang/protobuf v1.5.0 // indirect
+\tgithub.com/google/go-cmp v0.5.5 // indirect
+\tgolang.org/x/xerrors v0.0.0-20191204190536-9bdfabe68543 // indirect
+)
+"""
+
+GO_PROTOBUF_GO_SUM = """\
+github.com/golang/protobuf v1.5.0 h1:LUVKkCeviFUMKqHa4tXIIij/lbhnMbP7Fn5wKdKkRh4=
+github.com/golang/protobuf v1.5.0/go.mod h1:FsONVRAS9T7sI+LIUmWTfcYkHO4aIWwzhcaSAoJOfIk=
+github.com/google/go-cmp v0.5.5 h1:Khx7svrCpmxxtHBq5j2mp/xVjsi8hQMfNLvJFAlrGgU=
+github.com/google/go-cmp v0.5.5/go.mod h1:v8dTdLbMG2kIc/vJvl+f65V22dbkXbowE6jgT/gNBxE=
+golang.org/x/xerrors v0.0.0-20191204190536-9bdfabe68543 h1:E7g+9GITq07hpfrRu66IVDexMakfv52eLZ2CXBWiKr4=
+golang.org/x/xerrors v0.0.0-20191204190536-9bdfabe68543/go.mod h1:I/5z698sn9Ka8TeJc9MKroUUfqBBauWjQqLJ2OPfmY0=
+google.golang.org/grpc v1.2.0 h1:v8eFdETH8nqZHQ9x+0f2PLuU6W7zo5PFZuVEwH5126Y=
+google.golang.org/grpc v1.2.0/go.mod h1:yo6s7OP7yaDglbqo1J04qKzAhqBH6lvTonzMVmEdcZw=
+google.golang.org/grpc/cmd/protoc-gen-go-grpc v1.2.0 h1:TLkBREm4nIsEcexnCjgQd5GQWaHcqMzwQV0TX9pq8S0=
+google.golang.org/grpc/cmd/protoc-gen-go-grpc v1.2.0/go.mod h1:DNq5QpG7LJqD2AamLZ7zvKE0DEpVl2BSEVjFycAAjRY=
+google.golang.org/protobuf v1.27.1 h1:SnqbnDw1V7RiZcXPx5MEeqPv2s79L9i7BJUlG/+RurQ=
+google.golang.org/protobuf v1.27.1/go.mod h1:9q0QmTI4eRPtz6boOQmLYwt+qCgq0jsYwAQnmE0givc=
+"""
+
+
+@rule
+async def setup_go_protoc_plugin(platform: Platform) -> SetupGoProtocPlugin:
+    go_mod_digest = await Get(
+        Digest,
+        CreateDigest(
+            [
+                FileContent("go.mod", GO_PROTOBUF_GO_MOD.encode()),
+                FileContent("go.sum", GO_PROTOBUF_GO_SUM.encode()),
+            ]
+        ),
+    )
+
+    download_sources_result = await Get(
+        ProcessResult,
+        GoSdkProcess(
+            ["mod", "download", "all"],
+            input_digest=go_mod_digest,
+            output_directories=("gopath",),
+            description="Download Go `protoc` plugin sources.",
+            allow_downloads=True,
+        ),
+    )
+
+    go_plugin_build_result, go_grpc_plugin_build_result = await MultiGet(
+        Get(
+            ProcessResult,
+            GoSdkProcess(
+                ["install", "google.golang.org/protobuf/cmd/protoc-gen-go@v1.27.1"],
+                input_digest=download_sources_result.output_digest,
+                output_files=["gopath/bin/protoc-gen-go"],
+                description="Build Go protobuf plugin for `protoc`.",
+                env={"__GO_PROTOBUF_PLATFORM": str(platform)},
```

Comment:
@Eric-Arellano: This is how I tried to mix in the platform into cache key. Not generic but should be good enough for now.

Reply:
Instead, this should set `platform=platform` in the underlying `Process()` constructor to leverage the engine's MultiPlatform support. See https://github.com/pantsbuild/pants/pull/12725